Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -523,34 +522,74 @@ private func createResolvedPackages(
-    // Use a list to track all the module aliases specified for a given target to handle an override later
-    var nameToAliasAndPkgMap = [String: [(String, String)]]()
-    packageBuilders.forEach { $0.package.targets.forEach { $0.dependencies.forEach { dep in
-        if case let .product(prodRef, _) = dep {
-            if let prodPkg = prodRef.package {
-                if let prodModuleAliases = prodRef.moduleAliases {
-                    for (depName, depAlias) in prodModuleAliases {
-                        // Add a module alias to a map if specified
-                        nameToAliasAndPkgMap[depName, default: []].append((depAlias, prodPkg))
+// Create a map between a package and module aliases specified for the targets in the package.
+private func gatherModuleAliases(for rootPkgID: PackageIdentity?,
+                                 with packagesByIdentity: [PackageIdentity: ResolvedPackageBuilder]) -> [PackageIdentity: [String: String]] {
+    var result = [PackageIdentity: [String: String]]()
+    
+    // There could be multiple root packages but the common cases involve
+    // just one root package; handling multiple roots is tracked rdar://88518683
+    if let rootPkg = rootPkgID {
+        var pkgStack = [PackageIdentity]()
+        walkPkgTreeAndGetModuleAliases(for: rootPkg, with: packagesByIdentity, using: &pkgStack, output: &result)
+    }
+    return result
+}
+
+// Walk a package dependency tree and set the module aliases for targets in each package.
+// If multiple aliases are specified in upstream packages, aliases specified most downstream
+// will be used.
+private func walkPkgTreeAndGetModuleAliases(for pkgID: PackageIdentity,
+                                             with packagesByIdentity: [PackageIdentity: ResolvedPackageBuilder],
+                                             using pkgStack: inout [PackageIdentity],
+                                             output result: inout [PackageIdentity: [String: String]]) {
+    // Get the builder first
+    if let builder = packagesByIdentity[pkgID] {
+        // Add the pkgID to a stack used to keep track of
+        // multiple aliases specified in the package chain
+        pkgStack.append(pkgID)
+        builder.package.targets.forEach { target in
+            target.dependencies.forEach { dep in
+                // Check if a dependency for this target has module aliases specified
+                if case let .product(prodRef, _) = dep {
+                    if let prodPkg = prodRef.package {
+                        if let prodModuleAliases = prodRef.moduleAliases {
+                            for (depName, depAlias) in prodModuleAliases {
+                                let prodPkgID = PackageIdentity.plain(prodPkg)
+                                // Add the specified alias and the dependency package to a map
+                                result[prodPkgID, default: [:]][depName] = depAlias
```

Comment:
Is it possible for multiple product references to specify aliases for the same `depName`? If yes, I think they would override each other undetected here which might be something we want to explicitly diagnose.